Index: elfutils-0.155/libdwfl/linux-kernel-modules.c
===================================================================
--- elfutils-0.155.orig/libdwfl/linux-kernel-modules.c
+++ elfutils-0.155/libdwfl/linux-kernel-modules.c
@@ -601,8 +601,8 @@ check_module_notes (Dwfl_Module *mod)
 int
 dwfl_linux_kernel_report_kernel (Dwfl *dwfl)
 {
-  Dwarf_Addr start;
-  Dwarf_Addr end;
+  Dwarf_Addr start = 0;
+  Dwarf_Addr end = 0;
   inline Dwfl_Module *report (void)
     {
       return INTUSE(dwfl_report_module) (dwfl, KERNEL_MODNAME, start, end);
