From: Stephan Kulow <coolo@novell.com>

Avoid crash if the header is 0

Index: elfutils-0.155/libdw/dwarf_getcfi_elf.c
===================================================================
--- elfutils-0.155.orig/libdw/dwarf_getcfi_elf.c
+++ elfutils-0.155/libdw/dwarf_getcfi_elf.c
@@ -76,7 +76,7 @@ parse_eh_frame_hdr (const uint8_t *hdr,
 {
   const uint8_t *h = hdr;
 
-  if (*h++ != 1)		/* version */
+  if (!h || *h++ != 1)		/* version */
     return (void *) -1l;
 
   uint8_t eh_frame_ptr_encoding = *h++;
