--- elfutils-0.158/tests/run-strip-reloc.sh.0017~	2014-03-05 14:57:15.019418994 +0100
+++ elfutils-0.158/tests/run-strip-reloc.sh	2014-03-05 14:59:10.039420228 +0100
@@ -105,6 +105,6 @@ runtest hello_aarch64.ko 1
 
 # self test, shouldn't impact non-ET_REL files at all.
 runtest ${abs_top_builddir}/src/strip 0
-runtest ${abs_top_builddir}/src/strip.o 1
+runtest ${abs_top_builddir}/src/strip.o 0
 
 exit $status
