From: unknown
Upstream: no
Subject: do not build po and tests components

Index: elfutils-0.155/Makefile.am
===================================================================
--- elfutils-0.155.orig/Makefile.am
+++ elfutils-0.155/Makefile.am
@@ -24,7 +24,7 @@ pkginclude_HEADERS = version.h
 
 # Add doc back when we have some real content.
 SUBDIRS = config m4 lib libelf libebl libdwfl libdw libcpu libasm backends \
-	  src po tests
+	  src
 
 EXTRA_DIST = elfutils.spec GPG-KEY NOTES CONTRIBUTING \
 	     COPYING COPYING-GPLV2 COPYING-LGPLV3
Index: elfutils-0.155/Makefile.in
===================================================================
--- elfutils-0.155.orig/Makefile.in
+++ elfutils-0.155/Makefile.in
@@ -259,7 +259,7 @@ pkginclude_HEADERS = version.h
 
 # Add doc back when we have some real content.
 SUBDIRS = config m4 lib libelf libebl libdwfl libdw libcpu libasm backends \
-	  src po tests
+	  src
 
 EXTRA_DIST = elfutils.spec GPG-KEY NOTES CONTRIBUTING \
 	     COPYING COPYING-GPLV2 COPYING-LGPLV3
